### PR TITLE
MAPED data merging with Torch

### PR DESCRIPTION
**Pull request by henrygbell**
# What does this PR do?
This PR is the torchified version of [PR-169](https://github.com/electronmicroscopy/quantem/pull/169). For my data it speeds the merge_datasets method up by ~10x or more and adds batching.

# Example notebook
[maped_test_nb_torch.ipynb](https://github.com/user-attachments/files/26449944/maped_test_nb_torch.ipynb)

# Data for that notebook
Download [here](https://drive.google.com/drive/folders/1CPl6Z0rKKleaQnG2Nkhj3ix8bZmh35nL?usp=drive_link)

# For reviewers:
Please test the code on your own datasets! Check the code for accuracy. Let me know if you have any suggestions

TODO next:
- Can we optimize this torch version?
- dscan alignment

### CODE REVIEW

**Comment by bobleesj** [comment]
Happy to review and also address the remaining bottleneck. @henrygbell

File: src/quantem/diffraction/maped.py [hunk 1]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
 
-def shift_images(
-    images,
-    shifts_rc,
-    edge_blend: float = 8.0,
-    padding=None,
-    pad_val: str | float = 0.0,
-    shift_method: str = "bilinear",
-):
+class MAPEDTorch(AutoSerialize):
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py:815]: general comment - we need an end-to-end test either with synthetic data or data that used before refactoring w/ torch to ensure nothing broke in this PR. Then I would be more comfortable reviewing this.

**Comment by bobleesj** [comment]
@gvarnavi seems like CI tests aren't running on branch PRs, could we also add automated workflows? I want to ensure that this torch refactoring doesn't break the existing maped pytest.

File: src/quantem/diffraction/maped.py [hunk 2]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
+        weight_scale: float = 1 / 8,
+        plot_aligned: bool = True,
+        **plot_kwargs: Any,
+    ) -> MAPED:
+        """
+        Align mean diffraction patterns using weighted cross-correlation in Fourier space.
+
+        Parameters
+        ----------
+        edge_blend
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: pls follow numpy docstring - parameter type missing.

Comment (bobleesj) [reply src/quantem/diffraction/maped.py]: important for doc api doc rendering automated

File: src/quantem/diffraction/maped.py [hunk 3]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
+        plot_summary: bool = True,
+        scale: float | Sequence[float] | None = None,
+        **plot_kwargs: Any,
+    ) -> MAPED:
+        """
+        Compute dataset summary images.
+
+        Stores
+        ------
+        self.scales : torch.tensor
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: yeah here i see paramters

File: src/quantem/diffraction/maped.py (hunk 1 above)
Comment (bobleesj) [reply src/quantem/diffraction/maped.py:815]: second comment is that did the API or is it just torching without any user behavior modified? Please indicate that so that we don't need to worry about backward compatability in your PR comment

File: src/quantem/diffraction/maped.py [hunk 4]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
 
-    Returns
-    -------
-    np.ndarray
-        Blended image of shape (H + 2*padding, W + 2*padding).
+    This class manages a set of 4D-STEM datasets and provides utilities to:
+    - compute mean BF and mean DP summaries,
+    - choose/find diffraction origins,
+    - align diffraction space and real space,
+    - merge datasets into a single composite Dataset4dstem.
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py:823]: good

File: src/quantem/diffraction/maped.py [hunk 5]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
+        **plot_kwargs: Any,
+    ) -> MAPED:
+        """
+        Choose or automatically find the origin in diffraction space.
 
-            F = np.fft.fft2(stack[ind])
-            stack[ind] = np.fft.ifft2(F * ramp).real
+        Parameters
+        ----------
+        origins
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: numpy doc, same comment

File: src/quantem/diffraction/maped.py [hunk 6]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
-                prefilter=False,
-            )
-            stack_w[ind] = ndi_shift(
-                stack_w[ind],
-                shift=(shifts_rc[ind, 0], shifts_rc[ind, 1]),
-                order=1,
-                mode="constant",
-                cval=0.0,
-                prefilter=False,
+        Stores
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: stores or returns?

File: src/quantem/diffraction/maped.py [hunk 7]
```diff
@@ -761,132 +784,1370 @@ def merge_datasets(
...
+        1D Tukey window of length N.
+    """
+    n = torch.arange(N, device=device, dtype=dtype)
+    w = torch.ones(N, device=device, dtype=dtype)
+
+    if alpha <= 0:
+        return w
+    if alpha >= 1:
+        return torch.hann_window(N, device=device, dtype=dtype)
+
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py:1909]: just code comment - no extra whitespace needed, fewer lines better in general so that we don't have to scroll too much. I

**Review by bobleesj — COMMENTED** [review]
@henrygbell I spent 15-20 mins reviewing. Before I run with real data, I noticed that new utility torch functions were added. It is not required to write unit tests for each (if you want, better if you compare against scipy/numpy) but at least we need to know that we maintain parity after refactoring. Please see the drift PR below where torching was done but no API/accuracy was changed

Ref: https://github.com/electronmicroscopy/quantem/pull/206/changes#r3040892988. 

After your feedback, I will test with the data we collected at snsf.

File: src/quantem/diffraction/maped.py [hunk 8]
```diff
@@ -761,132 +813,1423 @@ def merge_datasets(
...
+        edge_w_dp = 1.0 - coverage
+
+        # Determine batch size (somewhat arbitrary)
+        if batch_size is None:
+            batch_size = max(1, min(32, Rout // 2))
+
+        c_out = torch.arange(Cout, dtype=torch.float32, device=self.device)
+        c_base = c_out - real_space_padding
+
+        merged = torch.zeros((Rout, Cout, Hp, Wp), dtype=torch.float64, device=self.device)
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py:1569]: reason using float64 instead of 32? I wonder 10^-5+ precision is important here.

**Review by bobleesj — COMMENTED** [review]
Quick memory management comment after today's meeting.

File: src/quantem/diffraction/maped.py [hunk 9]
```diff
@@ -761,132 +813,1423 @@ def merge_datasets(
...
+                    warnings.warn(
+                        f"Output overflow for dtype {dtype_out}: data range [{below}, {above}] exceeds "
+                        f"[{dmin}, {dmax}]. Values will be clipped.",
+                        RuntimeWarning,
+                    )
+                merged_out = torch.rint(torch.clamp(merged_f, dmin, dmax)).to(dtype=dtype_out)
+        else:
+            merged_out = merged.to(dtype=dtype_out)
+
+        dataset_merged = Dataset4dstem.from_array(array=merged_out.cpu().numpy())
```
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: We will fix this via https://github.com/electronmicroscopy/quantem/pull/228/

Comment (bobleesj) [reply src/quantem/diffraction/maped.py]: But aftrer this PR merged.

**Review by bobleesj — COMMENTED** [review]
Another low-hanging but important one.

File: src/quantem/diffraction/maped.py
Comment (bobleesj) [inline src/quantem/diffraction/maped.py]: @henrygbell we can fix this now. no numpy init needed. use `from_tensor()` to reduce memory footprint.

**Review by bobleesj — COMMENTED** [review]
you can probably fix this now

**Comment by bobleesj** [comment]
@henrygbell seems like we are having some PR synching issues here as well. 

A few things to check

1. `upstream/namoeam` is getting old, it is diverging from `upstream/dev`. Why don't you just make MAPED code merged to `upstream/dev` after making `upstream/nanobeam` merged `to `upstream/dev`? Please chat with @cophus 

This `upstream/nanobeam' is getting a bit too old where problems are starting to occur 

https://github.com/electronmicroscopy/quantem/pull/235#issuecomment-4608224270
